src/throttler/throttler.ts

```diff
@@ -254,4 +239,71 @@ export abstract class Throttler<T, R> {
       return p.resolve();
     });
   }
+
+  private initializeTimeout(cursorIndex: number): Promise<void> {
+    const taskData = this.taskDataMap.get(cursorIndex)!;
+    const timeoutMillis = taskData.timeoutMillis!;
+    const timeoutPromise = new Promise<void>((_, reject) => {
+      taskData.timeoutId = setTimeout(() => {
+        taskData.isTimedOut = true;
+        reject(new TimeoutError(this.taskName(cursorIndex), timeoutMillis));
+      }, timeoutMillis);
+    });
+
+    return timeoutPromise;
+  }
+
+  private async executeTask(cursorIndex: number): Promise<any> {
+    const taskData = this.taskDataMap.get(cursorIndex)!;
+    const t0 = Date.now();
+    try {
+      const result = await this.handler(taskData.task);
```

Comment:
We should check `taskData.isTimedOut` immediately after this `await` line.

Reply:
Alternatively, we could have different representation for these three stats:
1. Neither promise finished
2. Timeout won the race
3. Task won the race

There is no way to tell 1 and 3 apart right.
Have an enum with 3 states gives us strong guarantee of state transitions.